### PR TITLE
[task-manager] ♻️ Better support for fast refresh, improved typings

### LINKED ISSUE #6619: Location.startLocationUpdatesAsync() throws error "TaskManager.defineTask() not being called in global scope" after saving changes
<!-- IMPORTANT: Due to the holidays, please expect limited to no response from the Expo Team during the weeks of Christmas and New Year's (from 12/23 to 1/3). Additionally, issue catch-up will likely extend into the week of January 6. Thanks for your patience, we hope you enjoy the holidays, and have a happy New Year! 

### Environment

  Expo CLI 3.11.3 environment info:
    System:
      OS: Windows 10
    Binaries:
      Yarn: 1.19.1 - C:\Users\musta\scoop\apps\yarn\current\Yarn\bin\yarn.CMD
      npm: 6.9.0 - C:\Program Files\nodejs\npm.CMD
    IDEs:
      Android Studio: Version  3.5.0.0 AI-191.8026.42.35.6010548

Platform Android 9

### Steps to Reproduce

1. TaskManager.defineTask() in App.js outside of component class.
2. Register task with Location.startLocationUpdatesAsync() anywhere (normal or async function or lifecycle method ot useEffect hook).
3. Log gotten results inside Task.

### Expected Behavior

Background task should be started. On every location update there should be logged new locations array containg freshly caught location regardless of app being open or minimized.

### Actual Behavior

**TaskManager.defineTask must be called during initialization phase!** error is logged in the console. If Location.startLocationUpdatesAsync() is called in async function it also throws error in Client as well. Location is checked once or twice per app refresh.

### Reproducible Demo

```javascript
import React, { Component } from 'react';
import { Platform, Text, View, StyleSheet } from 'react-native';
import Constants from 'expo-constants';
import * as Location from 'expo-location';
import * as Permissions from 'expo-permissions';
import * as TaskManager from 'expo-task-manager';

export default class App extends Component {
  state = {
    location: null,
    errorMessage: null,
  };

  constructor() {
    super();
    this._ensurePermissionGranted();
    this._startLocationUpdates();
  }
  
  _startLocationUpdates = () => {
    Location.startLocationUpdatesAsync('watch', {
      foregroundService: {
        notificationTitle: 'Geotracker',
        notificationBody: 'Tracking enabled'
      }
    });
  }

  _ensurePermissionGranted = () => {
    if (Platform.OS === 'android' && !Constants.isDevice) {
      this.setState({
        errorMessage: 'Oops, this will not work on Sketch in an Android emulator. Try it on your device!',
      });
    }
  }

  render() {
    let text = 'Waiting..';
    if (this.state.errorMessage) {
      text = this.state.errorMessage;
    } else if (this.state.location) {
      text = JSON.stringify(this.state.location);
    }

    return (
      <View style={styles.container}>
        <Text style={styles.paragraph}>{text}</Text>
      </View>
    );
  }
}

TaskManager.defineTask('watch', ({ data: { locations = [] }, error }) => {
  if(error) {
    return console.error(error)
  }
  console.log(locations.reverse()[0]);
})

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    paddingTop: Constants.statusBarHeight,
    backgroundColor: '#ecf0f1',
  },
  paragraph: {
    margin: 24,
    fontSize: 18,
    textAlign: 'center',
  },
});
```
This code is as simple as it can get to show what exactly is not working. Whole project is fresh out of expo init command and code above is everything App.js consists of.

### PR DESCRIPTION
**Pull request by tsapeta**
# Why

Fixes #6619, improves TypeScript typings and adds doc comments.

# How

- Removed check whether the `defineTask` is called in top-level scope or not. That was causing unnecessary warnings and some problems with fast refresh or snacks. This check is not really needed now and appropriate warning will be logged either way but later, when the task is gonna be executed but there is no executor defined on the JS side.
- Exported some types that were not exported.
- Added doc comments.

# Test Plan

Tested in NCL example on iOS simulator while tracking device's location in the background and overriding tasks in the meantime by calling `defineTask` multiple times for the same task using fast refresh.

### CODE REVIEW

File: packages/expo-task-manager/src/TaskManager.ts [hunk 1]
```diff
@@ -133,9 +199,3 @@ if (ExpoTaskManager) {
     }
   );
 }
-
-// @tsapeta: Turn off `defineTask` function right after the initialization phase.
```
Comment (mczernek) [inline packages/expo-task-manager/src/TaskManager.ts]: Are we planning to replace this mechanism w some other one?

**Review by mczernek — APPROVED** [review]

File: packages/expo-task-manager/src/TaskManager.ts [hunk 2]
```diff
@@ -2,62 +2,109 @@ import { EventEmitter, UnavailabilityError } from '@unimodules/core';
...
-  if (tasks.has(taskName)) {
-    console.warn(`TaskManager.defineTask: task '${taskName}' is already defined.`);
-    return;
-  }
-  tasks.set(taskName, task);
+  tasks.set(taskName, taskExecutor);
 }
 
+/**
+ * Used to check whether the task is already defined.
```
Comment (bbarthec) [inline packages/expo-task-manager/src/TaskManager.ts]:
```suggestion
 * Checks whether the task is already defined.
```

File: packages/expo-task-manager/src/TaskManager.ts [hunk 3]
```diff
@@ -67,6 +114,11 @@ export async function isTaskRegisteredAsync(taskName: string): Promise<boolean>
   return ExpoTaskManager.isTaskRegisteredAsync(taskName);
 }
 
+/**
+ * Method returning an `options` object that the task was registered with.
```
Comment (bbarthec) [inline packages/expo-task-manager/src/TaskManager.ts]:
```suggestion
 * Retrieves an `options` object for provided `taskName`.
```

File: packages/expo-task-manager/src/TaskManager.ts [hunk 4]
```diff
@@ -76,14 +128,24 @@ export async function getTaskOptionsAsync<TaskOptions>(taskName: string): Promis
...
+export async function getRegisteredTasksAsync(): Promise<TaskManagerTask[]> {
   if (!ExpoTaskManager.getRegisteredTasksAsync) {
     throw new UnavailabilityError('TaskManager', 'getRegisteredTasksAsync');
   }
 
   return ExpoTaskManager.getRegisteredTasksAsync();
 }
 
+/**
+ * Method that unregisters the task. Tasks are usually registered by other modules (e.g. expo-location).
```
Comment (bbarthec) [inline packages/expo-task-manager/src/TaskManager.ts]:
```suggestion
 * Unregisters the task. Tasks are usually registered by other modules (e.g. expo-location).
```

File: packages/expo-task-manager/src/TaskManager.ts [hunk 5]
```diff
@@ -2,62 +2,109 @@ import { EventEmitter, UnavailabilityError } from '@unimodules/core';
...
   };
 }
 
-export interface RegisteredTask {
+/**
+ * An interface of already registered task.
+ */
+export interface TaskManagerTask {
+  /**
+   * Name under which the task was registered.
```
Comment (bbarthec) [inline packages/expo-task-manager/src/TaskManager.ts]:
```suggestion
   * Name that the task is registered with.
```

File: packages/expo-task-manager/src/TaskManager.ts [hunk 6]
```diff
@@ -2,62 +2,109 @@ import { EventEmitter, UnavailabilityError } from '@unimodules/core';
...
+   */
   executionInfo: {
     eventId: string;
     taskName: string;
   };
 }
 
-export interface RegisteredTask {
+/**
+ * An interface of already registered task.
```
Comment (bbarthec) [inline packages/expo-task-manager/src/TaskManager.ts]: I don't think we need to write in a comment what's the semantics of given element 😉 
```suggestion
 * Represents an already registered task.
```

File: packages/expo-task-manager/src/TaskManager.ts [hunk 7]
```diff
@@ -2,62 +2,109 @@ import { EventEmitter, UnavailabilityError } from '@unimodules/core';
...
 }
 
-export function defineTask(taskName: string, task: Task) {
-  if (!isRunningDuringInitialization) {
-    console.error(`TaskManager.defineTask must be called during initialization phase!`);
-    return;
-  }
+/**
+ * Method that you use to define a task – it saves given task executor under given task name
+ * which must be correlated with the task name used when registering the task.
```
Comment (bbarthec) [inline packages/expo-task-manager/src/TaskManager.ts]: Your comment is too complicated IMO ... 😢 

Why not to use comment from our docs? 😊
```suggestion
 * Defines task function.
 * It must be called in the global scope of your JavaScript bundle. In particular,
 * it cannot be called in any of React lifecycle methods like `componentDidMount`.
 * This limitation is due to the fact that when the application is launched in the background,
 * we need to spin up your JavaScript app, run your task and then shut down — no views are 
 * mounted in this scenario.
```

Or enrich/enhance it even further 🤔 
```suggestion
 * Defines a task. Provided `taskExecutor` subscribes for the tasks that are registered under `taskName`.
 * It must be called in the global scope of your JavaScript bundle. In particular,
 * it cannot be called in any of React lifecycle methods like `componentDidMount`.
 * This limitation is due to the fact that when the application is launched in the background,
 * we need to spin up your JavaScript app, run your task and then shut down — no views are 
 * mounted in this scenario.
```

**Review by bbarthec — CHANGES_REQUESTED** [review]
I think it needs some more adjustments 😉

File: packages/expo-task-manager/src/TaskManager.ts (hunk 1 above)
Comment (tsapeta) [reply packages/expo-task-manager/src/TaskManager.ts]: I don't see any other way to do what we want here and that works well with fast refresh, so I just removed that and allowed users to redefine their tasks.

File: packages/expo-task-manager/src/TaskManager.ts (hunk 7 above)
Comment (tsapeta) [reply packages/expo-task-manager/src/TaskManager.ts]: I didn't want to copy things from the docs - imho our docs should be a place where you can get more details about something and here we should only give some short and basic explanations.

**Review by bbarthec — APPROVED** [review]

**Pull request merged**